File: frontend/src/app/core/services/admin-subscription.service.ts
import { Injectable, inject } from '@angular/core';
import { ApiService } from './api.service';
import { pageParams } from '../../shared/utils/page-params.util';
import {
  Subscription,
  SubscriptionDetail,
  SubscriptionKpis,
  SubscriptionFilters,
  PlanHistoryEntry,
  Page,
} from '../../models/admin-subscription.model';

/**
 * Platform-staff API client for Plans & Subscriptions: the subscriptions list (filterable +
 * paginated), KPI totals, per-org detail, plan-change, and plan-history. Reads need PLAN_VIEW/
 * BILLING_VIEW; plan changes need PLAN_MANAGE (enforced + audit-logged on the backend).
 * (Plan catalog CRUD is the existing AdminService; grant-credit is AdminQuotaService.)
 */
@Injectable({ providedIn: 'root' })
export class AdminSubscriptionService {
  private readonly api = inject(ApiService);
  private readonly base = '/admin/subscriptions';

  list(filters: SubscriptionFilters = {}, page = 0, size = 10) {
    const params = pageParams(filters, page, size);
    return this.api.get<Page<Subscription>>(this.base, { params });
  }

  kpis() {
    return this.api.get<SubscriptionKpis>(`${this.base}/kpis`);
  }

  get(orgId: string) {
    return this.api.get<SubscriptionDetail>(`${this.base}/${orgId}`);
  }

  planHistory(orgId: string) {
    return this.api.get<PlanHistoryEntry[]>(`${this.base}/${orgId}/plan-history`);
  }

  changePlan(orgId: string, planId: string, reason: string | null) {
    return this.api.patch<SubscriptionDetail>(`${this.base}/${orgId}/plan`, { planId, reason });
  }
}

File: backend/src/main/java/com/postwerk/service/AdminSubscriptionService.java
package com.postwerk.service;

import com.postwerk.dto.admin.AdminSubscriptionDetailResponse;
import com.postwerk.dto.admin.AdminSubscriptionResponse;
import com.postwerk.dto.admin.PlanHistoryEntryResponse;
import com.postwerk.dto.admin.SubscriptionKpisResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

/**
 * Platform-staff Plans &amp; Subscriptions: every organization's plan assignment + derived MRR + metered
 * AI usage, with plan-change (and links to the existing suspend/activate + grant-credit features).
 * Reads need {@code PLAN_VIEW}/{@code BILLING_VIEW}; plan changes need {@code PLAN_MANAGE} (enforced
 * at the controller).
 *
 * @since 1.0
 */
public interface AdminSubscriptionService {

    /**
     * Paginated + filterable subscriptions list (filter/sort/paginate in-memory over all orgs — the
     * usage filter is derived from the effective AI cap, which has no single column to query on).
     *
     * @param search matches org name / slug / owner name+email / plan (blank = all)
     * @param plan   plan name | null
     * @param status "active" | "suspended" | null
     * @param usage  "over90" | "unlimited" | "aiOff" | null
     */
    Page<AdminSubscriptionResponse> listSubscriptions(String search, String plan, String status,
                                                      String usage, Pageable pageable);

    SubscriptionKpisResponse kpis();

    AdminSubscriptionDetailResponse getSubscription(UUID orgId);

    AdminSubscriptionDetailResponse changePlan(UUID orgId, UUID planId, String reason, UUID actorUserId, String ip);

    List<PlanHistoryEntryResponse> planHistory(UUID orgId);
}
